### ISSUE TITLE
Beam Search Decoder API

### ISSUE THREAD

**Issue by eaplatanios** [tensorflow/tensorflow#15323]
@ebrevdo Why is it that the user needs to call `tile_batch` explicitly for beam search decoders when using attention models? Couldn't the beam search decoder internally tile the provided `initial_state` in its constructor? It seems that this API is prone to wrong usage so I'm trying to understand why it's necessary.

Thank you!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by ebrevdo** [comment]
This is a good question; and I unfortunately I did not document the entire
sequence of design decisions leading to this.  However, at some point we
saw it was necessary to either do this, or add a lot of special logic
inside the beam search decoder to add special casing for *specific*
RNNCells; and we did not want to do this.

On Tue, Dec 12, 2017 at 1:45 PM, Anthony Platanios <notifications@github.com

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by reedwm** [comment]
Closing since the question has been answered.

**Comment by eaplatanios** [comment]
@ebrevdo I see. That's what I thought. I'm wondering if, in general, the RNN cells API could be redesigned to make such extensions easier. Are you guys considering that at all?

**Comment by eaplatanios** [comment]
@ebrevdo FYI, I revisited this for my Scala API after all this time, because it felt quite awkward to have to do this and I realized that there is a very simple design change that resolves the issue and allows tiling to be handled within the beam search decoder. All that is needed is for the memory and the memory sequence lengths tensors to be part of the attention wrapper RNN cell, rather than provided at construction time. It's a simple change but has resolved the issue for me and cannot see why it would not resolve it for the Python implementation too.

**Comment by ebrevdo** [comment]
Thanks for the suggestion Anthony, do you have an example diff in your
codebase showing the change to the attention wrapper rnn cell?

**Comment by eaplatanios** [comment]
@ebrevdo The actual commit is [this](https://github.com/eaplatanios/tensorflow_scala/commit/8efc1069d0a29ae92faf9a5eb3864c92ffbacc3b), however, it might be hard to follow if you are not familiar with Scala. In summary, the changes were the following:

- When constructing attention mechanisms, no memory needs to be provided, as shown in [this diff](https://github.com/eaplatanios/tensorflow_scala/commit/8efc1069d0a29ae92faf9a5eb3864c92ffbacc3b#diff-ef76330f5dbe0574036286de5f6430f8). In order to construct the initial attention state, a batch size needs to be provided and in order to compute an alignment, some previous attention state needs to be provided, that contains the keys, values, previous attention state, and optionally sequence lengths. It is the job of the attention wrapper cell to provide those, as described next.
- When constructed, the attention wrapper cell takes as input pairs of memories, and attention mechanisms (instead of just attention mechanisms), as shown in [this diff](https://github.com/eaplatanios/tensorflow_scala/commit/8efc1069d0a29ae92faf9a5eb3864c92ffbacc3b#diff-e5b5a14ee310ebfe2ede73ae49f8412e). These memories are used to initialize the attention states, as shown [here](https://github.com/eaplatanios/tensorflow_scala/commit/8efc1069d0a29ae92faf9a5eb3864c92ffbacc3b#diff-e5b5a14ee310ebfe2ede73ae49f8412eR115), which are part of the RNN cell state and can thus be tiled automatically by the beam search decoder.

Note that for the tiling I do recursion on the type structures of the cell state, at compile-time, but something similar could be done in Python using the TF nest module.

Does this help? Please let me know and if needed I could provide more detail. :)

**Comment by ebrevdo** [comment]
Thanks for this suggestion!  I will see if we have some cycles to try this
out.  We may try to make this change in the transition to TF 2.0; where we
can make backwards breaking changes in the API.

+scott, james

**Comment by ebrevdo** [comment]
+thang

**Comment by eaplatanios** [comment]
@ebrevdo Nice, thanks! It's just a suggestion...I'm not using that API since I use TF Scala, but I thought it might be useful to others and in simplifying the Python API. I'm currently attending a conference, but I can generally answer any questions and help once I get back. :)
